File: touchbar-recorder/TouchbarRecorder.m
//
//  TouchbarRecorder.m
//  touchbar-recorder
//
//  Created by Christoph Leimbrock on 20.02.18.
//  Copyright © 2018 Christoph Leimbrock. All rights reserved.
//

#import "TouchbarRecorder.h"
#import <QuartzCore/QuartzCore.h>
#import <AVFoundation/AVFoundation.h>
#import "DFR.h"
#import "SurfaceRecorder.h"

@interface TouchbarRecorder () <NSMachPortDelegate>
@property CGDisplayStreamRef stream;
@property dispatch_queue_t queue;
@property SurfaceRecorder *recorder;
@end

@implementation TouchbarRecorder

- (instancetype)init
{
    if (self = [super init]) {
    }
    
    return self;
}

- (void)start {
    _isRecording = YES;
    
    self.queue = dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0) ;
    dispatch_async(self.queue, ^{
        [self createDisplayStream];
    
        self.recorder = [SurfaceRecorder new];
        [self.recorder setupForSize:[DFR GetScreenSize]];
        
        if(self.stream) {
            CGDisplayStreamStart(self.stream);
        }
    });
}


- (void)createDisplayStream {
    self.stream = [DFR DisplayStreamCreate:0 queue:self.queue handler:^(CGDisplayStreamFrameStatus status, uint64_t displayTime, IOSurfaceRef  _Nullable surface, CGDisplayStreamUpdateRef  _Nullable updateRef) {
        if(!self.isRecording) return;
        if(!surface) {
            return;
        }
        
        if(status != kCGDisplayStreamFrameStatusFrameComplete) {
            return;
        }
        

        [self.recorder recordSurface:surface at:displayTime];
    }];
}

- (void)stop {
    dispatch_sync(self.queue, ^{
        if(self.stream) {
            CGDisplayStreamStop(self.stream);
            self.stream = NULL;
        }
        
        dispatch_semaphore_t sem = dispatch_semaphore_create(0);
        dispatch_block_t block = ^{
            dispatch_semaphore_signal(sem);
        };
        [self.recorder.writer finishWritingWithCompletionHandler:block];
        
        _isRecording = NO;
        while(dispatch_semaphore_wait(sem, dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)))) ;
    });
}

- (void)writeTo:(NSURL*)url {
    [[NSFileManager defaultManager] moveItemAtURL:self.recorder.outputFile toURL:url error:nil];
    [self.recorder cleanup];
}
@end

File: touchbar-recorder/TouchBarRecorder.swift
//
//  TouchBarRecorder_Swift.swift
//  touchbar-recorder
//
//  Created by Christoph Leimbrock on 21.02.18.
//  Copyright © 2018 Christoph Leimbrock. All rights reserved.
//

import Foundation
import AVFoundation

class TouchbarRecorder {
    var isRecording: Bool = false
    private var stream: CGDisplayStream?
    private var queue: DispatchQueue
    private var recorder: SurfaceRecorder
    
    init() {
        queue = DispatchQueue.global(qos: .userInitiated)
        recorder = SurfaceRecorder()
    }
    
    public func start() {
        self.isRecording = true;
        
        self.queue.async {
            self.stream = self.createStream()
            guard self.stream != nil else { return }
            self.recorder.setup(for: DFR.getScreenSize())
            self.stream!.start()
        }
    }
    
    private func createStream() -> CGDisplayStream? {
        if let stream = DFR.displayStreamCreate(0, queue: self.queue, handler: { (status, time, surface, _) in
            guard self.isRecording else { return }
            guard surface != nil else { return }
            guard status == .frameComplete else { return }
            
            self.recorder.record(surface, at: time)
        }) {
            return stream.takeUnretainedValue()
        }
        
        return nil
    }
    
    public func stop() {
        self.queue.async {
            if let stream = self.stream {
                stream.stop()
                self.stream = nil
            }
            
            let semaphore = DispatchSemaphore(value: 0)
            self.recorder.writer.finishWriting {
                semaphore.signal()
            }
            semaphore.wait()
            
            self.isRecording = false
        }
    }
    
    public func write(to url: URL) {
        try? FileManager.default.moveItem(at: self.recorder.outputFile, to: url)
        self.recorder.cleanup()
    }
}
